### PR TITLE
Support d-spacing workspaces in IntegratePeaksSkew and improve peak validation

### LINKED ISSUE #35271: Add additional peak validation options to IntegratePeaksSkew
**Is your feature request related to a problem? Please describe.**
Peak mask validation seems to be letting through too many spurious peaks and in particualr overestiomates I/sigma of weak peaks (due to the window getting quite small)

**Describe the solution you'd like**

1. Only maximise I/sigma of peaks with initial I/sigma above a threshold ~ 10
2. Allow window to move but stay width determined from B2Bexp params  for 5 < I/sigma < 10?
3. Set minimum window to be fraction of initial window (previously have set min number TOF bins)

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**

Several improvements to the `IntegratePeaksSkew` algorithm

1. Allow integration of workspaces in d-spacing (previously just TOF) - this required re-writing a lot of code to use mantid algorithms as opposed to numpy operations (due to non-aligned bin edges)
2. Add option not to optimise 1D window size below a threshold Intensity/sigma ratio
3. Improve formatting of output plot by putting legend outside the axes
4. Validate peaks based on number of statistically significant non-zero TOF bins

In addition a bug in estimating resolution parameters has been fixed

Note that as part of this PR this PR the error on empty bins is no longer increased to 1 (as is done in SXD2001) as this has been show to be invalid in issue #34841

**To test:**

(1) Run this script

```
Load(Filename='SXD26062.raw', OutputWorkspace='SXD26062')

CreatePeaksWorkspace(InstrumentWorkspace='SXD26062', NUmberOfPeaks=0, OutputWorkspace='SingleCrystalPeakTable')

AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=2366.8220873404448, DetectorID=34389, Height=1772, BinCount=95.229762433928016)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=2126.2022543400831, DetectorID=33804, Height=1505, BinCount=27.480506255931004)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=2184.0008770008217, DetectorID=35949, Height=1808, BinCount=52.661280309925587)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=1995.2115486292446, DetectorID=31765, Height=1513, BinCount=23.65730722975853)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=2252.1712809128194, DetectorID=30058, Height=2302, BinCount=84.449848622374006)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=5843.6050001340082, DetectorID=15121, Height=1639, BinCount=63.362132138439677)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=4419.0059861014552, DetectorID=38933, Height=885, BinCount=17.337073210576683)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=3023.5648640387335, DetectorID=37810, Height=710, BinCount=12.82320474223436)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=5145.6860467522974, DetectorID=38691, Height=1325, BinCount=126.70521154333336)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=4975.3450154521515, DetectorID=8558, Height=1345, BinCount=7.0523493183511672)
AddPeak(PeaksWorkspace='SingleCrystalPeakTable', RunWorkspace='SXD26062', TOF=16259.3, DetectorID=19474, Height=39, BinCount=6.5623281756867184)
 
ConvertUnits(InputWorkspace='SXD26062', Target='dSpacing',  OutputWorkspace='SXD26062')

fpath = '/mnt/babylon/Public/RWaite/SXD26062_peaks_int_skew.pdf'
IntegratePeaksSkew(InputWorkspace='SXD26062', PeaksWorkspace='SingleCrystalPeakTable', 
                   # NCols=3, NRows=3, NRowMax=3,NColMax=3,
                   BackscatteringTOFResolution=0.06, ThetaWidth=1.5, 
                   ScaleThetaWidthByWavelength=True, OptimiseMask=False, NVacanciesMax=1, 
                   UseNearestPeak=True, UpdatePeakPosition=False, 
                   OutputFile=fpath, 
                   LorentzCorrection=False, OutputWorkspace='SXD26062_peaks_int', 
                   OptimiseXWindowSize=True,
                   ThresholdIOverSigma=0.0)
```
(2) Inspect the output table - the peaks should be integrated
(3) Inspect the output pdf - the peak window should look like it covers the majority of the peaks (in 2D and 1D)
Fixes #35271

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py [hunk 1]
```diff
@@ -295,22 +346,24 @@ def integrate_peak(
...
                     self.status = new_status
                     # refocus data and re-optimise TOF limits
+                    self.x_integrated_data = opt_xintegrated_data
                     self.peak_mask = opt_peak_mask
                     self.non_bg_mask = np.logical_and(self.non_bg_mask, opt_non_bg_mask)
                     self.focus_data_in_detector_mask()
-                    self.find_peak_limits(min(self.ixlo, self.ixlo_opt), max(self.ixhi, self.ixhi_opt))
-            if self.ixhi_opt - self.ixlo_opt < min_nbins:
+                    self.find_peak_limits(self.ixmin_opt, self.ixmax_opt, optimise_xwindow, threshold_i_over_sig)
+            if np.sum(self.ypk[self.ixmin_opt : self.ixmax_opt] / np.sqrt(self.epk_sq[self.ixmin_opt : self.ixmax_opt]) > 1) < min_nbins:
```
Comment (jhaigh0) [inline Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:354]: Don't quite understand the final `> 1` in lhs of this expression. My understanding is that `np.sum` takes an array like input, but it looks like it's `np.sum(array > 1)`. Is that shorthand for a filter maybe?

**Review by jhaigh0 — CHANGES_REQUESTED** [review]
Good changes. Using the existing algorithms helps to simplify things and means there's hopefully less to test. I can't claim to fully understand all the maths, but at the very least it all made sense (no unused variables or min / maxes the wrong way round etc).

The pdf output looked good to me but less certain on how to check the `peaks_int` data. Here's a screenshot of what I got back and hopefully this looks correct.
image

Just a couple of minor other things.

**Comment by sf1919** [comment]
@RichardWaiteSTFC  As there are what appears to be valid test failures we won't re-run the ones that are currently 'scheduled' as they will rerun when you push a new commit

**Comment by RichardWaiteSTFC** [comment]
> @RichardWaiteSTFC As there are what appears to be valid test failures we won't re-run the ones that are currently 'scheduled' as they will rerun when you push a new commit

@sf1919 - thanks for checking, I believe these systems test failures are unrelated to this PR though so hopefully will pass when I next push!

File: docs/source/release/v6.7.0/Diffraction/Single_Crystal/Improvements/35411.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+- Improved formatting of plots in pdf output of :ref:`IntegratePeaksSkew <algm-IntegratePeaksSkew>`
+- Parameter ``NTOFBinsMin`` in :ref:`IntegratePeaksSkew <algm-IntegratePeaksSkew>` now refers to the number of non-zero bins in the integration window with y/error > 1
```
[the remark this answers is not in the document]
Comment (RichardWaiteSTFC) [reply docs/source/release/v6.7.0/Diffraction/Single_Crystal/Improvements/35411.rst:2]: OK - is there no improvements section now?

File: Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py (hunk 1 above)
Comment (RichardWaiteSTFC) [reply Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:354]: This is counting the number of bins with y/e > 1. Calling `sum` on a `bool` array casts each element to an `int` (`1` or `0`) - so it gives the number of `True` elements. I could store the result in a variable called something like `n_nonzero_bins` (by nonzero I mean statistically likely to not be zero given the error on the point)?

File: Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py [hunk 3]
```diff
@@ -210,53 +221,81 @@ def get_peak_region_array(self, peak, detid, bank_name, nrows, ncols, nrows_edge
...
-                signal[irow, icol, :] = self.ws.readY(ispec)
-                errors[irow, icol, :] = self.ws.readE(ispec)
-        # get x bin centers
-        xvals = self.ws.readX(int(ispecs[irow_peak, icol_peak]))
-        if len(xvals) > self.ws.blocksize():
-            xvals = 0.5 * (xvals[:-1] + xvals[1:])  # convert to bin centers
-        return xvals, signal, errors, irow_peak, icol_peak, detector_edges, detids
+        detids, det_edges, irow_peak, icol_peak = self.get_detid_array(bank, detid, row, col, drows, dcols, nrows_edge, ncols_edge)
+        peak_data = PeakData(irow_peak, icol_peak, det_edges, detids, peak, self.ws)
+        return peak_data
```
[the remark this answers is not in the document]
Comment (RichardWaiteSTFC) [reply Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:226]: > The pdf output looked good to me but less certain on how to check the `peaks_int` data. Here's a screenshot of what I got back and hopefully this looks correct. image
> 

Sorry I should have been more specific - can you check the peaks with a given HKL in the table have the same I/sigma as printed in the pdf plot titles? Thanks!

**Comment by RichardWaiteSTFC** [comment]
> The pdf output looked good to me but less certain on how to check the `peaks_int` data. Here's a screenshot of what I got back and hopefully this looks correct. image
>

Sorry I should have been more specific - can you check the peaks with a given HKL in the table have the same I/sigma as printed in the pdf plot titles? Thanks!

File: docs/source/release/v6.7.0/Diffraction/Single_Crystal/Improvements/35411.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (jhaigh0) [reply docs/source/release/v6.7.0/Diffraction/Single_Crystal/Improvements/35411.rst:2]: No there isn't. You can check by looking in `docs/release/templates/diffraction.rst`

**Comment by jhaigh0** [comment]
> > The pdf output looked good to me but less certain on how to check the `peaks_int` data. Here's a screenshot of what I got back and hopefully this looks correct. image
> 
> Sorry I should have been more specific - can you check the peaks with a given HKL in the table have the same I/sigma as printed in the pdf plot titles? Thanks!

Yes, they do match the values in the `SXD26062_peaks_int` 👍

**Review by jhaigh0 — APPROVED** [review]

File: docs/source/release/v6.7.0/Diffraction/Single_Crystal/Improvements/35411.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (sf1919) [reply docs/source/release/v6.7.0/Diffraction/Single_Crystal/Improvements/35411.rst]: For info Improvements section was remove when we changed to separate release notes as it was generally felt if it wasn't a bug fix it was a new feature - even if it extends the ability of an exisiting feature.

**Comment by sf1919** [comment]
@mantidproject/gatekeepers Adding screenshot to show all tests passed and PR approved
Screenshot (263)

**Comment by RichardWaiteSTFC** [comment]
@mantidproject/gatekeepers - I know you're probably busy but if someone is free could they look at this? Thanks!

File: Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py [hunk 4]
```diff
@@ -194,69 +205,92 @@ def _get_detid_array_rect_detector(self, bank, detid, row, col, drows, dcols, nr
...
-        self.det_edges = det_edges
-        self.dets = dets
-        # set initial integration limits
-        self.initial_frac_width = dTOF / self.tof
-        self.ixlo = np.argmin(abs(xpk - (self.tof - 0.5 * dTOF)))
-        self.ixhi = np.argmin(abs(xpk - (self.tof + 0.5 * dTOF)))
-        self.ixpk = np.argmin(abs(xpk - self.tof))
-        self.ixlo_opt, self.ixhi_opt = self.ixlo, self.ixhi
-        self.ypk, self.epk_sq = None, None
+        if self.ws.getAxis(0).getUnit().unitID() == "dSpacing":
```
Comment (thomashampson) [inline Framework/PythonInterface/plugins/algorithms/IntegratePeaksSkew.py:239]: Is there a more encapsulated way of checking this?

**Pull request merged**